### ISSUE TITLE
Sharded-axis `argmax` silently produces wrong results (lowers to `all_reduce<Sum>` on indices)

### ISSUE THREAD

**Issue by elogir** [tenstorrent/tt-mlir#8623]
## Summary

When `stablehlo.reduce` with argmax semantics is sharded over its reduce axis, the cross-shard reduction is materialized as a `ttnn.all_reduce`. TTIR's `AllReduceOp` rejects every reduction kind except `Sum`. When the indices output of argmax flows through that all_reduce, per-shard indices are summed across chips instead of being combined into a global argmax, silently producing wrong tokens.

## Source

https://github.com/tenstorrent/tt-mlir/blob/d18541ab8e7ecdae71c6e104bcb54ce2be9de378/lib/Dialect/TTIR/IR/TTIROps.cpp#L5485-L5487

## Reproducer

```python
import os
os.environ["CONVERT_SHLO_TO_SHARDY"] = "1"
import torch, torch.nn as nn, torch_xla
import torch_xla.runtime as xr
import torch_xla.distributed.spmd as xs

class LMHead(nn.Module):
    def __init__(self):
        super().__init__()
        self.w = nn.Parameter(torch.randn(1024, 8192, dtype=torch.bfloat16))
    def forward(self, x):
        return torch.argmax(torch.matmul(x, self.w), -1)

torch.manual_seed(0)
m = LMHead()
x = torch.randn(1, 1024, dtype=torch.bfloat16)
expected = torch.argmax(torch.matmul(x, m.w.detach()), -1).item()

xr.set_device_type("TT")
xr.use_spmd()
d = torch_xla.device()
n = xr.global_runtime_device_count()
mesh = xs.Mesh(list(range(n)), (n,), ("model",))

torch_xla.set_custom_compile_options(
    {"export_path": "ir", "export_model_name": "argmax_sharded_voc"})
m = m.to(d)
x = x.to(d)
xs.mark_sharding(m.w, mesh, (None, "model"))
out = m(x)
torch_xla.sync()
device_val = out.cpu().item()
print(f"device argmax: {device_val}")
print(f"cpu expected:  {expected}")
print(f"agree:         {device_val == expected}")
```

Output:

```
device argmax: 1818
cpu expected:  833
agree:         False
```

## Possible resolution

For greedy decoding the natural pattern can be to shard the vocab axis of `lm_head` and `argmax` over it. The result should be the correct global index, regardless of how the input was sharded. Today the only way to get a correct answer is to keep the reduced axis replicated, forcing an inverted TP layout for the rest of `lm_head`.

- Either: compose a correct cross-shard argmax
- Or at least: reject sharded-axis argmax with a clear compile-time error rather than silently lowering through a Sum-only collective.

Tested with `pjrt-plugin-tt==1.2.0.dev20260526002843`

**Comment by acicovicTT** [comment]
Looked into this a bit, here is the AI-generated summary of the initial investigation for this issue:

---

## Investigation summary

### Repro confirmed in both frontends
The bug reproduces identically in PyTorch/XLA and JAX (a JAX port of the
reproducer was written and run on n300, 2 chips):

| Frontend    | device argmax | cpu expected | sum of per-shard local indices (a+b) |
|-------------|---------------|--------------|--------------------------------------|
| PyTorch/XLA | 1818          | 833          | 1818                                 |
| JAX         | 192           | 50           | 192                                  |

In both cases `device == a + b`, i.e. the two per-shard local argmax indices are
summed instead of combined into a global argmax.

### The defect originates in the SPMD partitioner (Shardy), at StableHLO
Both frontends route through Shardy (`jax_use_shardy_partitioner=True` /
`CONVERT_SHLO_TO_SHARDY=1`). The wrong reduction is already present at StableHLO,
before any TTIR lowering. Post-partition graph shows a per-shard reduce followed
by a `stablehlo.all_reduce` whose reducer body is `stablehlo.add`, applied
to the index output only. TTIR's `all_reduce<sum>` is a faithful 1:1 lowering of
this; tt-mlir is not where the wrong result is introduced.

### Root cause: the cross-shard combine, not the iota
On a 1x2 mesh, per-shard local iota `0..4095` is correct and expected (each shard is 4096 wide).
The bug is that the cross-shard combine fails to (1) select the winning shard by
comparing per-shard max *values* and (2) add the winner's base offset
(`winner_shard * shard_width`). A correct result is
`global_idx = winner_shard*4096 + local_idx[winner]`. Shardy instead emits
`all_reduce<add>` over the local indices, producing `sum_s local_idx_s`, which is
meaningless. (Confirmed by the numbers: JAX expected 50 = winner in shard 0,
local index 50; device 192 = 50 + 142, summing an unrelated shard-1 local index.)

### argmax is not a composite in tt-mlir
argmax is recognized by structural pattern-matching on `stablehlo.reduce`
(`isArgMax` in StableHLOToTTIRPatterns.cpp, separate torch and JAX patterns) ->
`ttir.argmax` -> `ttnn.argmax`. It is absent from both composite registries
(StableHLOLegalizeCompositePass, TTNNResolveComposites). Because it arrives
decomposed (not wrapped in `stablehlo.composite`), Shardy partitions its internals
and mishandles the cross-shard index combine. Emitting argmax as a shardable
composite would let the partitioner treat it atomically.

### Proposed tt-mlir mitigation (turns silent miscompile into a clear error)
Detection signal at TTIR is clean: a `ttir.all_reduce` whose input is defined by
`ttir.argmax`. Implemented as a guard in `AllReduceOp::verify()` and
`AllReduceAsyncOp::verify()` (TTIROps.cpp):

    if (mlir::isa_and_nonnull<ttir::ArgMaxOp>(getInput().getDefiningOp()))
      return emitOpError("...mis-partitioned sharded-axis argmax...");

This is a symptom catcher for the canonical case, NOT a correctness fix.

### Known gap in the proposed solution above
A hand-written argmax-equivalent (`max`+`eq`+`where`+`iota`) lowers the index
extraction to `ttir.max` (not `ttir.argmax`) feeding the same `all_reduce<sum>`,
i.e. the same bug with no `ttir.argmax` in the graph. Built `ttmlir-opt` with the
guard and ran two cases:

- `ttir.argmax -> all_reduce<sum>`  => guard fires (error).
- `ttir.max    -> all_reduce<sum>`  => verifies cleanly, silent pass (gap).

Minimal principled broadening that closes the gap with low false-positive risk:
fire when the sum `all_reduce`'s input is defined by ANY non-sum reduction
(`ArgMaxOp` / `MaxOp` / `MinOp`), since summing per-shard max/min/argmax across
shards is never a valid global combine.

### Recommended resolution
1. Real fix is upstream: Shardy should combine sharded-axis argmax with
   value-selection + base-offset (or argmax should be a shardable composite so it
   is not decomposed). TODO: confirm whether this is a known Shardy/OpenXLA issue.
2. tt-mlir "fix": land the verifier guard so this fails to compile with a clear
   message instead of silently producing wrong tokens. Decide conservative
   (argmax-only) vs broadened (any non-sum reduction) scope given the gap above.

---

We will probably go with 1. proposed above to try and fix this in Shardy itself. That is yet TBD.
